Automattic/wp-calypso
# Jetpack connect: Rip out queryObject state

This PR separates the no direct access component from the authorization components.

It removes the page middleware responsible inserting `queryObject` in state.

It validates queryData against a schema and transforms it before passing the required props down to components as required.

It avoids a redirection which effectively strips the query from the URL.

It passes the `redirectAfterAuth` to the plans page via the query rather than state lookup.

Note that some analytics from the authorize components will not be fired when accessing the page directly. This is a slight change from the current behavior. I wouldn't expect that to be a problem cc/ @johnHackworth 

## Testing
1. Try accessing directly https://calypso.live/jetpack/connect/authorize?branch=update/jetpack-connect/auth-form-middleware
1. Ensure you see the no direct access warning 
1. Try a connection flow (logged-in/out) and ensure that it continues to work normally.

Flows:
- [x] Connect starting logged in
- [x] Connect starting logged out
- [x] WooCommerce Services
  * Edit Jetpack plugin to point to `calypso.localhost:3000` (P7rd6c-Me-p2)
  * Install WooCommerce Services plugin and start via the banner at WP Admin
  * Should auto-auth and skip plans
- [x] WooCommerce Wizard
  * Edit Jetpack plugin to point to `calypso.localhost:3000` (P7rd6c-Me-p2)
  * Install WooCommerce plugin, select an option using WooCommerce Services, and connect via banner
  * Should auto-auth and skip plans
- [x] Jetpack Partner NUX flow
  * Instructions at #20025 
- [x] JPO
  * Modify the "from" query param to a string starting with `jpo`
  * Should skip no plans
- [x] SSO (see https://github.com/Automattic/wp-calypso/pull/19956#issuecomment-347258158)
  1. Connect a site
  1. Enable SSO (activate default modules works)
  1. Add a new user with an email associated with an existing WordPress.com account
  1. While connected to WordPress.com the account from the previous step, visit the site `/wp-admin`
  1. Click the SSO button
  1. Make sure you hit the dev site
  1. Verify you're connected.
